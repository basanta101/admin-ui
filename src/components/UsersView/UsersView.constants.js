import { CheckBox, TableOptions, CellInput } from "../index";
import "./UsersView.css";

export const columns = ({
  handleCheckRow = (f) => f,
  handleCheckAllRow = (f) => f,
  isAllRowsInViewChecked = false,
  handleRowDelBtnClick = (f) => f,
  handleRowEditBtnClick = (f) => f,
  rowInEditModeData = {},
  onCellDataChange = (f) => f,
  onRowEditDone = (f) => f,
  selectedRows = [],
  onBlurCellInput = (f) => f,
  isAllRowsDeleted = false,
}) => {
  return [
    {
      Header: ({ data }) => {
        return (
          <CheckBox
            handleCheckCb={(checked) => handleCheckAllRow({ ...data, checked })}
            checked={isAllRowsInViewChecked}
            disabled={isAllRowsDeleted}
          />
        );
      },
      colId: "checkBoxCol",
      Cell: ({ data }) => {
        return (
          <CheckBox
            handleCheckCb={(checked) => handleCheckRow({ ...data, checked })}
            checked={selectedRows.includes(data.id)}
          />
        );
      },
    },
    {
      Header: () => <span className="colLabel">Name</span>,
      colId: "nameCol",
      Cell: ({ data }) => {
        const { id, name } = rowInEditModeData;
        const nameValue = data.name || "--";
        return (
          <>
            {id === data.id ? (
              <CellInput
                value={name}
                onChange={(val) => onCellDataChange({ type: "name", val })}
                pattern="[a-zA-Z]"
                onBlur={onBlurCellInput}
              />
            ) : (
              <span className="cellText" title={nameValue}>{nameValue}</span>
            )}
          </>
        );
      },
    },
    {
      Header: () => <span className="colLabel">Email</span>,
      colId: "emailCol",
      Cell: ({ data }) => {
        const { id, email } = rowInEditModeData;
        const emailValue = data.email || "--";
        return (
          <>
            {id === data.id ? (
              <CellInput
                value={email}
                onChange={(val) => onCellDataChange({ type: "email", val })}
                type="email"
                onBlur={onBlurCellInput}
              />
            ) : (
              <span className="cellText" title={emailValue}>
                {emailValue}
              </span>
            )}
          </>
        );
      },
    },
    {
      Header: () => <span className="colLabel">Role</span>,
      colId: "roleCol",
      Cell: ({ data }) => {
        const { id, role } = rowInEditModeData;
        const roleValue = data.role || "--";
        return (
          <>
            {id === data.id ? (
              <CellInput
                value={role}
                onChange={(val) => onCellDataChange({ type: "role", val })}
                onBlur={onBlurCellInput}
                autofocus
              />
            ) : (
              <span className="cellText" title={roleValue}>
                {roleValue}
              </span>
            )}
          </>
        );
      },
    },
    {
      Header: () => <span className="colLabel">Actions</span>,
      colId: "actionsCol",
      Cell: ({ data }) => {
        return (
          <TableOptions
            handleRowDelBtnClick={() => handleRowDelBtnClick(data)}
            handleRowEditBtnClick={() => handleRowEditBtnClick(data)}
            onDone={onRowEditDone}
            isRowInEditMode={data.id === rowInEditModeData.id}
          />
        );
      },
    },
  ];
};

export const pageSize = 10;

export const SEARCH_INPUT_PLACEHOLDER = "Search by name, email or role";
