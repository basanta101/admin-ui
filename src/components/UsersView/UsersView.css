.wrapper {
    padding: 16px;
    /* border: 1px solid blue; */
    height: calc(100% - 32px);
}

.footerWrapper {
    height: 50px;
    display: flex;
    justify-content: space-around;
    align-items: center;
}

.footerDelBtn {
    border: 1px solid #eee;
    background-color: #00d3ff;
    border-radius: 4px;
    color: rgba(0, 0, 0, 0.9);
    padding: 0 8px;
    height: 25px
}