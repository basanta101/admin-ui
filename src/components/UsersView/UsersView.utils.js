export const paginate = (array = [], pageSize = 10, pageNumber = 1) => {
  // human-readable page numbers usually start with 1, so we reduce 1 in the first argument
  return array.slice((pageNumber - 1) * pageSize, pageNumber * pageSize);
};

export const calculateTotalPages = (arr, pageSize) =>
  Math.ceil(arr.length / pageSize);

export const validateRowFields = (rowData = {}) => {
  let isValid = true;
  const errorKeys = [];
  let errMsg = "";
  const { id, ...fieldsToValidate } = rowData;
  Object.entries(fieldsToValidate).forEach(([key, value]) => {
    const isEmptyField = !value;
    const isInValidEmail = key === "email" && !/^\S+@\S+\.\S+$/.test(value);
    const isInValidRole =
      key === "role" && !["admin", "member"].includes(value.toLowerCase());
    const isInValidName = key === "name" && !/[a-zA-Z ]/.test(value);

    if (isEmptyField || isInValidEmail || isInValidRole || isInValidName) {
      errorKeys.push(key);
      isValid = false;
      return;
    }
  });

  errMsg = `
    Please Check: ${errorKeys.join(", ")} field(s).
    P.S: 
      Name cannot be empty, it can only contain aplhabets and spaces,
      Email cannot be invalid and Role can be either Member or Admin
  `;

  return { isValid, errorKeys, errMsg };
};
