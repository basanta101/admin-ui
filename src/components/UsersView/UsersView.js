import { useState, useEffect } from "react";
import _ from "lodash";
import {
  columns,
  pageSize,
  SEARCH_INPUT_PLACEHOLDER,
} from "./UsersView.constants";
import {
  paginate,
  calculateTotalPages,
  validateRowFields,
} from "./UsersView.utils";
import "./UsersView.css";
import { Pagination, Table, SearchInput } from "../index";
import { useFetch } from "../../utils";

const UsersView = () => {
  // replica of original data, kept for reference.
  const [originalUserData, updateOriginalUserData] = useState([]);
  /* The array (users), which will be modified, while performing table operations, 
    such as delete, edit
  */
  const [users, updateUsers] = useState([]);
  const [pagination, updatePagination] = useState({
    pageSize,
    currentPageNo: 1,
    totalPages: null,
    searchStartPageNo: 0,
  });
  const [searchText, updateSearchText] = useState({
    isDirty: false,
    value: "",
  });
  const [rowInEditModeData, updateRowInEditModeData] = useState({
    id: null,
    name: "",
    email: "",
    role: "",
  });
  const [selectedRows, updateSelectedRows] = useState([]); // rows which are checked.
  const [allRowsCheckedPages, updateAllRowsCheckedPages] = useState([]); // pages with all rows selected.

  useEffect(() => {
    // const url = `https://geektrust.s3-ap-southeast-1.amazonaws.com/adminui-problem/members.json`;
    // const { response: users } = useFetch(url);
    // const totalPages = calculateTotalPages(users, pageSize);
    // updateOriginalUserData(users);
    // updateUsers(users);
    // updatePagination({
    //   ...pagination,
    //   totalPages,
    // });
    try {
      fetch(
        `https://geektrust.s3-ap-southeast-1.amazonaws.com/adminui-problem/members.json`
      )
        .then((response) => response.json())
        .then((data) => {
          const totalPages = calculateTotalPages(data, pageSize);
          updateOriginalUserData(data);
          updateUsers(data);
          updatePagination({
            ...pagination,
            totalPages,
          });
        });
    } catch (err) {
      console.log("error while fetching users data", err);
    }
  }, []);

  const handlePageChange = (pageNo) => {
    const { pageSize } = pagination;
    const newTotalPages = calculateTotalPages(users, pageSize);
    if (pageNo === 0 || pageNo > newTotalPages) return;
    updatePagination({
      ...pagination,
      currentPageNo: pageNo,
      totalPages: newTotalPages,
    });
  };

  const handleSearchTextChange = (event) => {
    const { value } = event.target;
    const searchVal = value.toLowerCase();
    const { pageSize, currentPageNo, searchStartPageNo } = pagination;
    const filteredUsers = originalUserData.filter(
      (item) =>
        item.name.toLowerCase().includes(searchVal) ||
        item.role.toLowerCase().includes(searchVal) ||
        item.email.toLowerCase().includes(searchVal)
    );
    updateUsers(filteredUsers);
    const newTotalPages = calculateTotalPages(filteredUsers, pageSize);
    let newSearchStartPageNo = searchStartPageNo;
    if (!searchText.isDirty) {
      // isDirty is false, set the search start pageNo
      newSearchStartPageNo = currentPageNo;
    }
    updatePagination({
      ...pagination,
      currentPageNo: value ? 1 : searchStartPageNo, // if user searches
      totalPages: newTotalPages,
      searchStartPageNo: newSearchStartPageNo,
    });

    updateSearchText({ value, isDirty: value ? true : false });
  };

  const handleCheckRow = (checkedRowData) => {
    const { checked, id } = checkedRowData;
    // TODO: refactor to use one update
    if (checked && !selectedRows.includes(id)) {
      updateSelectedRows(selectedRows.concat(id));
    } else {
      updateSelectedRows(selectedRows.filter((item) => item !== id));
    }
  };

  const handleCheckAllRow = (checkedColData) => {
    const { checked } = checkedColData;
    let selectedIds = selectedRows;
    let allRowsInViewCheckedPageNos = allRowsCheckedPages;
    const { pageSize, currentPageNo } = pagination;
    const checkedUsers = paginate(users, pageSize, currentPageNo).map(
      (item) => item.id
    );
    if (checked) {
      // check all the rows in view, push their ids into selectedIds(array)
      selectedIds = selectedRows.concat(checkedUsers);
      // store the page no for which all the rows are selected.
      allRowsInViewCheckedPageNos =
        allRowsInViewCheckedPageNos.concat(currentPageNo);
    }

    if (!checked) {
      selectedIds = selectedRows.filter((row) => !checkedUsers.includes(row));
      allRowsInViewCheckedPageNos = allRowsInViewCheckedPageNos.filter(
        (pageNo) => pageNo !== currentPageNo
      );
    }

    updateSelectedRows(selectedIds);
    updateAllRowsCheckedPages(allRowsInViewCheckedPageNos);
  };

  const handleRowEditBtnClick = (data) => {
    updateRowInEditModeData({ ...data });
  };

  const onCellDataChange = ({ type, val }) => {
    updateRowInEditModeData({ ...rowInEditModeData, [type]: val });
  };

  const onRowEditDone = () => {
    // if any of the filed is empty alert the user and return
    const { isValid, errMsg } = validateRowFields(rowInEditModeData);
    if (!isValid) {
      alert(errMsg);
      return;
    }
    const editedUsers = users.map((item) =>
      item.id === rowInEditModeData.id ? { ...rowInEditModeData } : item
    );
    updateOriginalUserData([...editedUsers]);
    updateUsers(editedUsers);
    updateRowInEditModeData({
      id: null,
      name: "",
      email: "",
      role: "",
    });
  };

  const handleRowDelBtnClick = (data) => {
    const { totalPages: prevTotalPages, currentPageNo, pageSize } = pagination;
    const filteredUsers = users.filter((user) => user.id !== data.id);
    const newTotalPages = calculateTotalPages(filteredUsers, pageSize);
    updateUsers(filteredUsers);
    // let pagination
    // TODO: refactor to use one update
    if (newTotalPages < prevTotalPages) {
      updatePagination({
        ...pagination,
        currentPageNo: Math.max(1, currentPageNo - 1),
        totalPages: Math.max(1, newTotalPages),
      });
      return;
    }
    updatePagination({ ...pagination, totalPages: newTotalPages });
  };

  const handleFooterDelBtnClick = () => {
    if (!selectedRows.length) return; // no row is selected return.
    const { pageSize, currentPageNo } = pagination;
    let filteredUsers = users.filter((item) => !selectedRows.includes(item.id)); // users(rows of table) which are not selected
    const newTotalPages = calculateTotalPages(filteredUsers, pageSize);
    updateUsers(filteredUsers); // update the table by deleting selected users(table rows)
    updateAllRowsCheckedPages([]); // remove pageNo from all selected pageNos row arrays
    let newCurrentPageNo = currentPageNo;
    if (currentPageNo > newTotalPages) {
      // we need to decrement the current page no.
      newCurrentPageNo = newTotalPages;
    }
    updatePagination({
      ...pagination,
      currentPageNo: Math.max(1, newCurrentPageNo),
      totalPages: Math.max(1, newTotalPages),
    });
  };

  const getPaginatedData = (users) => {
    if (!users.length) return;
    const { currentPageNo, pageSize: currentPageSize } = pagination;
    return paginate(users, currentPageSize, currentPageNo);
  };

  const onBlurCellInput = () => {
    if (!rowInEditModeData.id) return;
    // if rowEdit was going on, save it.
    onRowEditDone();
  };

  return (
    <div className="wrapper">
      <SearchInput
        placeholder={SEARCH_INPUT_PLACEHOLDER}
        value={searchText.value}
        onChange={handleSearchTextChange}
      />
      <Table
        columns={columns({
          handleCheckRow,
          handleCheckAllRow,
          isAllRowsInViewChecked: allRowsCheckedPages.includes(
            pagination.currentPageNo
          ),
          handleRowDelBtnClick,
          handleRowEditBtnClick,
          rowInEditModeData,
          onCellDataChange,
          onRowEditDone,
          selectedRows,
          onBlurCellInput,
          isAllRowsDeleted: !users.length,
        })}
        data={getPaginatedData(users)}
        selectedRows={selectedRows}
      />
      <div className="footerWrapper">
        {users.length ? (
          <>
            <button
              className="footerDelBtn"
              onClick={handleFooterDelBtnClick}
              disabled={!selectedRows.length}
            >
              Delete Selected
            </button>
            <Pagination
              pagination={pagination}
              handlePageChange={handlePageChange}
            />
          </>
        ) : null}
      </div>
    </div>
  );
};

export default UsersView;

// TODO: check if any code is using spread, ans it requires a deep copy
// use lodash to do deep copy instead of spread operator.
