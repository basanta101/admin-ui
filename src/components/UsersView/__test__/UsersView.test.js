import { rest } from "msw";
import { setupServer } from "msw/node";
import render from '@testing-library/react';
import renderer from "react-test-renderer";
import UsersView from "../UsersView";

it("renders a snapshot of Pagination Component", () => {
  const tree = renderer.create(<UsersView />).toJSON();
  expect(tree).toMatchSnapshot();
});

const server = setupServer(
  rest.get(
    `https://geektrust.s3-ap-southeast-1.amazonaws.com/adminui-problem/members.json`
  ),
  async (req, res, ctx) => {
    return res(
      ctx.json([
        {
          id: 1,
          name: "Basanta Sharma",
          email: "bsbickysharma@gmail.com",
          role: "member",
        },
      ])
    );
  }
);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());
