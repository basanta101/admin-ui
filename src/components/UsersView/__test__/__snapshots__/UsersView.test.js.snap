// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a snapshot of Pagination Component 1`] = `
<div
  className="wrapper"
>
  <input
    className="searchInput"
    onChange={[Function]}
    placeholder="Search by name, email or role"
    type="text"
    value=""
  />
  <div
    className="tableWrapper"
  >
    <div
      className="columnWrapper"
    >
      <div
        className=""
      >
        <input
          checked={false}
          className="checkBox disabled"
          disabled={true}
          onChange={[Function]}
          type="checkbox"
        />
      </div>
      <div
        className="headerCol"
      >
        <span
          className="colLabel"
        >
          Name
        </span>
      </div>
      <div
        className="headerCol"
      >
        <span
          className="colLabel"
        >
          Email
        </span>
      </div>
      <div
        className="headerCol"
      >
        <span
          className="colLabel"
        >
          Role
        </span>
      </div>
      <div
        className="headerCol"
      >
        <span
          className="colLabel"
        >
          Actions
        </span>
      </div>
    </div>
    <div
      className="tableBody"
    />
  </div>
  <div
    className="footerWrapper"
  />
</div>
`;
